sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSchemaCapableIOProvider.java

```diff
@@ -0,0 +1,127 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io;
+
+import com.google.auto.service.AutoService;
+import java.io.Serializable;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.io.SchemaIO;
+import org.apache.beam.sdk.schemas.io.SchemaIOProvider;
+import org.apache.beam.sdk.schemas.transforms.Convert;
+import org.apache.beam.sdk.schemas.utils.AvroUtils;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.Row;
+
+/**
+ * An implementation of {@link SchemaIOProvider} for reading and writing Avro files with {@link
+ * AvroIO}.
+ */
+@Internal
+@AutoService(SchemaIOProvider.class)
+public class AvroSchemaCapableIOProvider implements SchemaIOProvider {
+  /** Returns an id that uniquely represents this IO. */
+  @Override
+  public String identifier() {
+    return "avro";
+  }
+
+  /**
+   * Returns the expected schema of the configuration object. Note this is distinct from the schema
+   * of the data source itself. No configuration expected for Avro.
+   */
+  @Override
+  public Schema configurationSchema() {
+    return Schema.builder().build();
+  }
+
+  /**
+   * Produce a SchemaIO given a String representing the data's location, the schema of the data that
+   * resides there, and some IO-specific configuration object.
+   */
+  @Override
+  public AvroSchemaIO from(String location, Row configuration, Schema dataSchema) {
+    return new AvroSchemaIO(location, dataSchema);
+  }
+
+  @Override
+  public boolean requiresDataSchema() {
+    return true;
+  }
+
+  @Override
+  public PCollection.IsBounded isBounded() {
+    return PCollection.IsBounded.BOUNDED;
+  }
+
+  /** An abstraction to create schema aware IOs. */
+  @Internal
+  private static class AvroSchemaIO implements SchemaIO, Serializable {
+    protected final Schema dataSchema;
+    protected final String location;
+
+    private AvroSchemaIO(String location, Schema dataSchema) {
+      this.dataSchema = dataSchema;
+      this.location = location;
+    }
+
+    @Override
+    public Schema schema() {
+      return dataSchema;
+    }
+
+    @Override
+    public PTransform<PBegin, PCollection<Row>> buildReader() {
+      return new PTransform<PBegin, PCollection<Row>>() {
+        @Override
+        public PCollection<Row> expand(PBegin begin) {
+          return begin
+              .apply(
+                  "AvroIORead",
+                  AvroIO.readGenericRecords(AvroUtils.toAvroSchema(dataSchema, null, null))
```

Comment:
This PR changes the behavior of this line, from setting the second parameter from table name to null. How will this affect the behavior of AvroIO?

Reply:
We changed the table name to null here because the general SchemaCapableIOTableProviderWrapper does not pass the table.getName() value to AvroSchemaCapableIOProvider. Since the toAvroSchema(..) method here has a nullable name parameter that is used to create a record, it didn't seem imperative to maintaining functionality. @TheNeuralBit What do you think?

Reply:
I think it is fine, when the name of the schema isn't specified we just call it "topLevelRecord": https://github.com/apache/beam/blob/f36250f2816da2a0d8350e92f0010615f8491ee0/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java#L327

I don't think the name has any effect on the way we encode/decode the Avro records. (CC @iemejia in case I'm wrong)

Reply:
Oups catching up on old stuff, AFAIR in theory it does but in practice it does not, because this is a part of the spec that is specified but nobody enforces specially the Java impl, and if the Java impl does not the others don't for compatibility reasons.